airflow/dag/serialization.py

```diff
@@ -0,0 +1,250 @@
+            new_x[k] = _serialize(v, visited_dags)
+    return new_x
+
+
+def _serialize_dag(x, visited_dags):
+    """Serialize a DAG."""
+    if x.dag_id in visited_dags:
+        return {TYPE: DAG, VAR: str(x.dag_id)}
+
+    new_x = {TYPE: DAG}
+    visited_dags[x.dag_id] = new_x
+    new_x[VAR] = _serialize_object(
+        x, visited_dags, included_fields=_dag_included_fields)
+    return new_x
+
+
+def _serialize_operator(x, visited_dags):
+    """Serialize an operator."""
+    return _encode(
+        _serialize_object(
+            x, visited_dags, included_fields=_op_included_fields),
+        type_=OP,
+        class_=x.__class__.__name__
+    )
+
+
+def _encode(x, type_, class_=None):
+    """Encode data by a JSON dict."""
+    return ({VAR: x, TYPE: type_} if class_ is None
+            else {VAR: x, TYPE: type_, CLASS: class_})
+
+
+def _serialize(x, visited_dags):  # pylint: disable=too-many-return-statements
+    """Helper function of depth first search for serialization.
+
+    visited_dags stores DAGs that are being stringifying for have been stringified,
+    for:
+      (1) preventing deadlock loop caused by task.dag, task._dag, and dag.parent_dag;
+      (2) replacing the fields in (1) with serialized counterparts.
+
+    The serialization protocol is:
+      (1) keeping JSON supported types: primitives, dict, list;
+      (2) encoding other types as {TYPE, 'foo', VAR, 'bar'}, the deserialization
+          step decode VAR according to TYPE;
+      (3) Operator has a special field CLASS to record the original class
+          name for displaying in UI.
+    """
+    try:
+        if _is_primitive(x):
+            return x
+        elif isinstance(x, dict):
+            return _encode({k: _serialize(v, visited_dags) for k, v in x.items()}, type_=DICT)
+        elif isinstance(x, list):
+            return [_serialize(v, visited_dags) for v in x]
+        elif isinstance(x, models.DAG):
+            return _serialize_dag(x, visited_dags)
+        elif isinstance(x, models.BaseOperator):
+            return _serialize_operator(x, visited_dags)
+        elif isinstance(x, _datetime_types):
+            return _encode(x.isoformat(), type_=DATETIME)
+        elif isinstance(x, datetime.timedelta):
+            return _encode(x.total_seconds(), type_=TIMEDELTA)
+        elif isinstance(x, pendulum.tz.Timezone):
+            return _encode(str(x.name), type_=TIMEZONE)
+        elif callable(x):
+            return str(get_python_source(x))
+        elif isinstance(x, set):
+            return _encode([_serialize(v, visited_dags) for v in x], type_=SET)
+        elif isinstance(x, tuple):
+            return _encode([_serialize(v, visited_dags) for v in x], type_=TUPLE)
+        else:
+            logging.debug('Cast type %s to str in serialization.', type(x))
+            return str(x)
+    except Exception:  # pylint: disable=broad-except
+        logging.warning('Failed to stringify.', exc_info=True)
+        return FAILED
```

Comment:
What happens if I have `MyOperator(some_var='failed')`, will that get confused for an actual error ?

Reply:
make it more specific: failed => serialization_failed

We should make sure "fail" never happen in normal use case. It is a guard that the whole recursion won't fail if one field is invalid.

Reply:
Perhaps `airflow__serialization_failed` then?